plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/ResourceContainer.java

```diff
@@ -0,0 +1,30 @@
+package org.jboss.tools.openshift.internal.ui.models;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.openshift.restclient.model.IResource;
+
+public abstract class ResourceContainer<R extends IResource, P extends IOpenshiftUIElement<?>>
+		extends AbstractResourceWrapper<R, P> {
+	protected Map<IResource, AbstractResourceWrapper<?, ?>> containedResources = new HashMap<>();
+	protected Object childrenLock = new Object();
+
+	public ResourceContainer(P parent, R resource) {
+		super(parent, resource);
+	}
+
+	public Collection<AbstractResourceWrapper<?, ?>> getResourcesOfKind(String kind) {

```

Comment:
Yes, the new model looks very good, the goal is certainly achieved. 
In determining if the method is called with low or high frequency, the model listens to changes to the remote project, they may be multiple on a single user action,model will update on each remote change, then refreshing a branch in explorer on each change event, UI will request for children for all nodes in the branch (maybe I miss something here in my reasoning). All this may be harmless for small models, and maybe optimization would only add unneeded complexity here. However, to be sure of it, I would advise to add a performance test for measuring time with models of, say, 10, 100, and 1000 resources, and computing the dependency of time on size. If it turns to be cubic growth or faster, the test should always fail, if it is square growth, the guess of real world model size should be taken into account to decide whether to fail or not.